# -----------------------------------------------------------------------------
# Copyright (c) 2019, Battelle National Biodefense Institute.
#
# This file is part of MicroHapDB (http://github.com/bioforensics/microhapdb)
# and is licensed under the BSD license: see LICENSE.txt.
# -----------------------------------------------------------------------------

from utils import smartopen, linköping_marker_coords, linköping_allele_frequencies
import pandas


rule all:
    input: 'marker.tsv', 'population.tsv', 'frequency.tsv'


rule dbsnp_subset:
    input:
        markers='marker-variants.tsv',
        dbsnp=config['dbsnp38'],
        rsidx=config['dbsnp38'].replace('.vcf.gz', '.rsidx'),
    output: 'dbsnp-subset.vcf.gz'
    run:
        markers = pandas.read_csv(input.markers, sep='\t')
        rsids = ' '.join(markers['rs-number'].tolist())
        command = f'rsidx search --header --out {output[0]} {input.dbsnp} {input.rsidx} {rsids}'
        shell(command)


rule marker_coords:
    input:
        script='./compile_marker_definitions.py',
        markers='marker-variants.tsv',
        dbsnp='dbsnp-subset.vcf.gz',
    output: 'marker.tsv'
    shell: '{input} {output}'


rule frequency:
    input:
        script='./compile_frequencies.py',
        markers='marker.tsv',
        freqs='allele-frequencies.tsv',
    output: 'frequency.tsv'
    shell: '{input} {output}'
